ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/OciConnectorExecutor.java

```diff
@@ -0,0 +1,189 @@
+@ConnectorExecutor(OCI_SIGV1)
+public class OciConnectorExecutor implements RemoteConnectorExecutor {
+
+    @Getter
+    private final OciConnector connector;
+
+    @Setter
+    @Getter
+    private ScriptService scriptService;
+
+    @Setter
+    @Getter
+    private TokenBucket rateLimiter;
+
+    @Setter
+    @Getter
+    private Map<String, TokenBucket> userRateLimiterMap;
+
+    @Setter
+    @Getter
+    private org.opensearch.client.Client client;
+
+    private final RestClient restClient;
+
+    public OciConnectorExecutor(Connector connector) {
+        this.connector = (OciConnector) connector;
+
+        final BasicAuthenticationDetailsProvider provider = buildAuthenticationDetailsProvider(connector);
+        final RestClientFactory restClientFactory = RestClientFactoryBuilder.builder().build();
+        final RequestSigner requestSigner = DefaultRequestSigner.createRequestSigner(provider);
+        this.restClient = restClientFactory.create(requestSigner, Collections.emptyMap());
+    }
+
+    @Override
+    public void invokeRemoteModel(MLInput mlInput, Map<String, String> parameters, String payload, List<ModelTensors> tensorOutputs) {
+        try {
+            final String endpoint = connector.getPredictEndpoint(parameters);
+            final String method = connector.getPredictHttpMethod();
+            final Response response = makeHttpCall(endpoint, method, payload);
+
+            final String modelResponse = getInputStreamContent((InputStream) response.getEntity());
+            final int statusCode = response.getStatus();
+            if (statusCode < 200 || statusCode >= 300) {
+                throw new OpenSearchStatusException(REMOTE_SERVICE_ERROR + modelResponse, RestStatus.fromCode(statusCode));
+            }
+
+            final ModelTensors tensors = processOutput(modelResponse, connector, scriptService, parameters);
+            tensors.setStatusCode(statusCode);
+            tensorOutputs.add(tensors);
+        } catch (IOException e) {
+            throw new MLException("Fail to execute predict in oci connector", e);
+        }
+    }
+
+    private Response makeHttpCall(String endpoint, String httpMethod, String payload) {
+        final WebTarget target = getWebTarget(endpoint);
+        final WrappedInvocationBuilder wrappedIb = new WrappedInvocationBuilder(target.request(), target.getUri());
+        final Response response;
+        switch (httpMethod.toUpperCase(Locale.ROOT)) {
+            case "POST":
+                response = restClient.post(wrappedIb, payload, new BmcRequest<>());
+                break;
+            case "GET":
+                response = restClient.get(wrappedIb, new BmcRequest<>());
+                break;
+            default:
+                throw new IllegalArgumentException("unsupported http method");
+        }
+        return response;
+    }
+
+    /**
+     *
+     * RestClient is a general wrapper of {@link Client} to talk to OCI services however it has a poor design
```

Comment:
RestClient is a general wrapper for the {@link Client}, designed to communicate with OCI services. However, it suffers from a poor design, especially regarding support for calling multiple endpoints. Specifically, when RestClient::setEndpoint is called, it creates a new instance of WebTarget with that endpoint. But instead of returning this instance directly, it requires the customer to make an additional call to access the WebTarget instance.

For now, let's make the calls to setEndpoint and getBaseTarget synchronized to avoid race conditions. Moreover, this approach is relatively inexpensive compared to making remote endpoint calls.

Reply:
Done